src/db.rs

```diff
@@ -14,43 +14,138 @@
 //
 
 use crate::{
+    column_family::AsColumnFamilyRef,
+    column_family::BoundColumnFamily,
     ffi,
     ffi_util::{from_cstr, opt_bytes_to_ptr, raw_data, to_cpath},
-    ColumnFamily, ColumnFamilyDescriptor, CompactOptions, DBIterator, DBPinnableSlice,
-    DBRawIterator, DBWALIterator, Direction, Error, FlushOptions, IngestExternalFileOptions,
-    IteratorMode, Options, ReadOptions, Snapshot, WriteBatch, WriteOptions,
-    DEFAULT_COLUMN_FAMILY_NAME,
+    ColumnFamily, ColumnFamilyDescriptor, CompactOptions, DBIteratorWithThreadMode,
+    DBPinnableSlice, DBRawIteratorWithThreadMode, DBWALIterator, Direction, Error, FlushOptions,
+    IngestExternalFileOptions, IteratorMode, Options, ReadOptions, SnapshotWithThreadMode,
+    WriteBatch, WriteOptions, DEFAULT_COLUMN_FAMILY_NAME,
 };
 
 use libc::{self, c_char, c_int, c_uchar, c_void, size_t};
 use std::collections::BTreeMap;
 use std::ffi::{CStr, CString};
 use std::fmt;
 use std::fs;
+use std::marker::PhantomData;
 use std::path::Path;
 use std::path::PathBuf;
 use std::ptr;
 use std::slice;
 use std::str;
+use std::sync::RwLock;
 use std::time::Duration;
 
+// Marker trait to specify single or multi threaded column family alterations for DB
+// Also, this is minimum common API sharable between SingleThreaded and
+// MultiThreaded. Others differ in self mutability and return type.
+pub trait ThreadMode {
+    fn new(cf_map: BTreeMap<String, ColumnFamily>) -> Self;
+    unsafe fn cf_drop_all(&mut self);
+}
+
+pub struct SingleThreaded {
```

Comment:
Description for **pub** struct ?

Reply:
done 2cfe746ae0dbe5a40c92c0e61f925bddd1f59714